service/storage_service.cc

```diff
@@ -1881,6 +1881,27 @@ future<> storage_service::join_token_ring(cdc::generation_service& cdc_gen_servi
     auto advertise = gms::advertise_myself(!replacing_a_node_with_same_ip);
     co_await _gossiper.start_gossiping(generation_number, app_states, advertise);
 
+    if (should_bootstrap()) {
+        // Wait for NORMAL state handlers to finish for existing nodes now, so that connection dropping
+        // (happening at the end of `handle_state_normal`: `notify_joined`) doesn't interrupt
+        // group 0 joining or repair. (See #12764, #12956, #12972, #13302)
+        //
+        // But before we can do that, we must make sure that gossip sees at least one other node
+        // and fetches the list of peers from it; otherwise `wait_for_normal_state_handled_on_boot`
+        // may trivially finish without waiting for anyone.
+        co_await wait_for_live_node_to_show_up();
+
+        auto ignore_nodes = ri
+                ? parse_node_list(_db.local().get_config().ignore_dead_nodes_for_replace(), get_token_metadata())
+                // TODO: specify ignore_nodes for bootstrap
+                : std::unordered_set<gms::inet_address>{};
+        auto sync_nodes = co_await get_nodes_to_sync_with(ignore_nodes);
+        if (ri) {
+            sync_nodes.erase(ri->address);
+        }
+        co_await wait_for_normal_state_handled_on_boot(sync_nodes);
+    }
```

Comment:
So I wonder what do we plan to do in "topology over raft" world. The only reason we have node states (only normal and left) in gossiper at all is for shutdown notifications to work. But the plan is to fix the shutdown notification to not rely on it (may be move it outside of gossiper) and drop the hack. Now this code will prevent bootstrapping when this refactoring  will happens. It will wait for gossiper's normal handler to be called and it will never be.

BTW right now with topology over raft we never even call `notify_joined`, so this is a bug (that is covered by "call notifier" item in the release plan). And when we fully remove node's state handling from gossiper the `notify_joined()` will be probably called from raft state machine which will happen after snapshot transfer which means first we will not drop the connection in the middle of it and second we may transfer a snapshot over a connection with wrong parameters.

Reply:
> So I wonder what do we plan to do in "topology over raft" world. The only reason we have node states (only normal and left) in gossiper at all is for shutdown notifications to work. But the plan is to fix the shutdown notification to not rely on it (may be move it outside of gossiper) and drop the hack.

I say, let's cross that bridge when we get to it.

"waiting for NORMAL" should be completely unnecessary in topology-over-raft world, which does explicit barriers whenever they are needed, so it will probably be enough to put `if (!topology_raft_enabled)` over this code or something similar.

> we may transfer a snapshot over a connection with wrong parameters.

wrong in this case doesn't mean unsafe though, it means suboptimal (e.g. we encrypt the connection when it's not necessary) so it's OK.

Reply:
> > So I wonder what do we plan to do in "topology over raft" world. The only reason we have node states (only normal and left) in gossiper at all is for shutdown notifications to work. But the plan is to fix the shutdown notification to not rely on it (may be move it outside of gossiper) and drop the hack.
> 
> I say, let's cross that bridge when we get to it.
> 
> "waiting for NORMAL" should be completely unnecessary in topology-over-raft world, which does explicit barriers whenever they are needed, so it will probably be enough to put `if (!topology_raft_enabled)` over this code or something similar.

Can we do it now? Not that I disagree with crossing the bridge part, but I am sure we will forget about it and will be surprised later.

> 
> > we may transfer a snapshot over a connection with wrong parameters.
> 
> wrong in this case doesn't mean unsafe though, it means suboptimal (e.g. we encrypt the connection when it's not necessary) so it's OK.

Yes.

Reply:
> Can we do it now?

OK, will do it in v2 (but let's wait for other reviews first)

Reply:
Am I getting this right that if you queue https://github.com/scylladb/scylladb/pull/14039 this patch won't be needed?

Reply:
I have an idea to simplify the shutdown message handling to get rid of the hack.

PR: https://github.com/scylladb/scylladb/pull/14424

Reply:
> Am I getting this right that if you queue https://github.com/scylladb/scylladb/pull/14039 this patch won't be needed?

For tests running in Raft mode maybe

Reply:
But this could be useful if we wanted to also disable the sleeps in dtests, for example

Reply:
> I have an idea to simplify the shutdown message handling to get rid of the hack.
> 
> PR: #14424

How will this help though? The reason the hack exists is because to send the shutdown RPC the node have to exist in the gossiper. I do not see that this pull request changes it.

Reply:
This is how the hack avoids sending the shutdown message in the decommission node:

- It sets the node status to LEFT which is one of the SILENT_SHUTDOWN_STATES
- So the check is_silent_shutdown_state returns true and the shutdown message is skipped.
```
if (my_ep_state && !is_silent_shutdown_state(*my_ep_state)) {
     _messaging.send_gossip_shutdown()
}
```

With the PR, it is always safe to send shutdown message since it does not change the status of the node but only mark the node who sends shutdown message as DOWN.

Reply:
Ah, OK. But I still need to add a node in NORMAL state so that shutdown message will be sent when node exits normally. And since we add it with NORMAL state the state needs to be removed when the node is removed/decommissioned. How can I do it without moving node to a LEFT state?

Reply:
In gossiper::do_stop_gossiping, a shutdown message is sent to any peer nodes that is in _live_endpoints list. There is no requirement that the local node has to be in NORMAL status.

Reply:
> In gossiper::do_stop_gossiping, a shutdown message is sent to any peer nodes that is in _live_endpoints list. There is no requirement that the local node has to be in NORMAL status.

Yes. I misremembered the reason. If there is no ep state for the node that is been shutdown the node does not send anything:

```
        if (my_ep_state && !is_silent_shutdown_state(*my_ep_state)) {
            auto local_generation = my_ep_state->get_heart_beat_state().get_generation();
            logger.info("Announcing shutdown");
            add_local_application_state(application_state::STATUS, versioned_value::shutdown(true)).get();
            auto live_endpoints = _live_endpoints;
            for (inet_address addr : live_endpoints) {
                msg_addr id = get_msg_addr(addr);
                logger.info("Sending a GossipShutdown to {} with generation {}", id.addr, local_generation);
                _messaging.send_gossip_shutdown(id, get_broadcast_address(), local_generation.value()).then_wrapped([id] (auto&&f) {
                    try {
                        f.get();
                        logger.trace("Got GossipShutdown Reply");
                    } catch (...) {
                        logger.warn("Fail to send GossipShutdown to {}: {}", id, std::current_exception());
                    }
                    return make_ready_future<>();
                }).get();
            }
            sleep(std::chrono::milliseconds(_gcfg.shutdown_announce_ms)).get();
        } else {
         if (my_ep_state && !is_silent_shutdown_state(*my_ep_state)) {
            auto local_generation = my_ep_state->get_heart_beat_state().get_generation();
            logger.info("Announcing shutdown");
            add_local_application_state(application_state::STATUS, versioned_value::shutdown(true)).get();
            auto live_endpoints = _live_endpoints;
            for (inet_address addr : live_endpoints) {
                msg_addr id = get_msg_addr(addr);
                logger.info("Sending a GossipShutdown to {} with generation {}", id.addr, local_generation);
                _messaging.send_gossip_shutdown(id, get_broadcast_address(), local_generation.value()).then_wrapped([id] (auto&&f) {
                    try {
                        f.get();
                        logger.trace("Got GossipShutdown Reply");
                    } catch (...) {
                        logger.warn("Fail to send GossipShutdown to {}: {}", id, std::current_exception());
                    }
                    return make_ready_future<>();
                }).get();
            }
            sleep(std::chrono::milliseconds(_gcfg.shutdown_announce_ms)).get();
        } else {
            logger.warn("No local state or state is in silent shutdown, not announcing shutdown");
        }
           logger.warn("No local state or state is in silent shutdown, not announcing shutdown");
        }
```

Reply:
@asias @gleb-cloudius isn't that a discussion for https://github.com/scylladb/scylladb/pull/14424?

Reply:
> @asias @gleb-cloudius isn't that a discussion for #14424?

May be. This is definitely not a discussion that should prevent the PR from been merged.

Reply:
if there are no peer nodes in gossip, gossip would not know who the shutdown message should be sent to? Do have endpoint_state  for the local host at all?

Reply:
> if there are no peer nodes in gossip, gossip would not know who the shutdown message should be sent to? Do have endpoint_state for the local host at all?

We still distribute some data through the gossiper, so there should be one.

Reply:
Then, the node should be able to send the shutdown messages.

Reply:
> Then, the node should be able to send the shutdown messages.

No, because of `!is_silent_shutdown_state(*my_ep_state)` check. Since there is no state the function the function returns true because `STATUS_UNKNOWN` is listed in `SILENT_SHUTDOWN_STATES`.

Reply:
So, NORMAL status is never inserted? Perhaps we can insert a STATUS_RAFT to indicate that the gossip node status is not used.

Reply:
> So, NORMAL status is never inserted?
With topology over raft we should non redistribute node's state through the gossiper. We do it through raft. But I had to add NORMAL state to gossiper for shutdown to work.

> Perhaps we can insert a STATUS_RAFT to indicate that the gossip node status is not used.

May be. Will it break anything?